### DISCUSSION TITLE
Drawback of sessions on by default

### DISCUSSION THREAD

**Discussion by jdtsmith** [emacs-lsp/lsp-mode#3204] [Ideas]
I thought I'd point out one drawback of the "auto-collect into sessions" approach lsp-mode takes. As a new user of lsp-mode, after a few hours of use, I found it caused emacs to grind to a halt.  I've since discovered what was really happening: I'd opened a python file in my _home directory_.  My lang server `pyright` happily tried to scan my entire (large) home directory, and it began choking.  I killed Emacs and restarted.  That should fix it!  But lsp-mode had quietly saved my home directory as part of a session in the file `~/.emacs.d/.lsp-session-v1`.  So even after killing emacs and pyright, my install was effectively "broken" from there on, since (as I didn't know at the time) lsp-mode consults this file and starts any new server looking through everything referenced in it (i.e. all files in my home-dir).  I uninstalled lsp-mode and moved on.

Now arguably this was a dumb move, but I bet it happens a lot.  I suppose the option I was supposed to have picked was blacklist or do nothing, but neither of those made sense at the time, and when I tried them, no server got run.   I think a default configuration of "we'll start a lang server for this project for you but won't remember it" might be simpler for new users, and prevent these kinds of issues.  I.e. opt-in to the idea of persistent sessions, so a "polluted" session list doesn't give non-obvious issues. 

A perhaps similar idea is a catchall "if I open a file anywhere random that would require importing into the session, consider it part of the following default project" to prevent over-eager language servers from scanning everything in sight.

**Comment by yyoncho** [comment]
> I think a default configuration of "we'll start a lang server for this project for you but won't remember it" might be simpler for new users, and prevent these kinds of issues

This will prevent some issues, but it will cause tons of other issues related to multiroot projects, and also not persisting the project roots will cause usability issue if we ask about project roots on each restart. I have plans to add a warning if you add home folder as a project root. I believe doom has that in their implementation of this feature. 

> A perhaps similar idea is a catchall "if I open a file anywhere random that would require importing into the session, consider it part of the following default project" to prevent over-eager language servers from scanning everything in sight.

I have plans to implement that as opt-in option in order to support root-less mode.

**Reply by jdtsmith** [reply]
Great thanks.  Root-less mode sounds ideal.  I'm actually surprised language servers don't have this, some sensible "I don't know where I am" backup, perhaps via a general global config file or something.  Maybe some of them do.

**Reply by yyoncho** [reply]
Initially, a lot of servers were crashing when started without root. After that it came as a scenario for handling standalone files.

**Reply by jdtsmith** [reply]
I think non-programmers tend to have more "standalone, one-off" files that don't merit setting up a full workspace/virtual environment (for example).  Handling this kind of file gracefully seems essential (for me anyway).
